### PR TITLE
[1650] Create Apply draft trainees from ApplyApplications

### PR DESCRIPTION
**Pull request by archferns**
### Context
https://trello.com/c/ZMG2nW2v/1650-m-import-application-service

### Changes proposed in this pull request
Based on https://github.com/DFE-Digital/register-trainee-teachers/pull/566, add a service to process apply_applications and have them converted into a draft trainee.

### Guidance to review
1. Create an apply application
2. Call the `Trainees::CreateFromApply` service with the apply application
3. Expect to have a new trainee created with the info held in the apply application

```ruby
apply_application = FactoryBot.create(:apply_application)
Trainees::CreateFromApply.call(application: apply_application)
```

### CODE REVIEW

File: app/services/degrees/create_from_apply.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        grade: attributes["grade"],
+      }
+    end
+
+    def non_uk_degree_params
+      {
+        locale_code: Trainee.locale_codes[:non_uk],
+        # TODO: This should be the comparable_uk_degree` from Apply
+        non_uk_degree: attributes["non_uk_qualification_type"],
+        country: country,
```
Comment (archferns) [inline app/services/degrees/create_from_apply.rb:56]: Worth noting that this information could be unavailable for non-degree qualifications 
https://github.com/DFE-Digital/apply-for-teacher-training/blob/ea87f086b9e55a21248fb53143c1f36689b64735/spec/presenters/hesa_qualification_fields_presenter_spec.rb#L33-L46

However, would be worth double checking with apply as the API docs have no mention of any `hesa_*` fields against qualifications https://sandbox.apply-for-teacher-training.service.gov.uk/register-api#get-applications

**Comment by AbigailMcP** [comment]
[[M] Import application service](https://trello.com/c/ZMG2nW2v/1650-m-import-application-service)

File: app/services/trainees/create_from_apply.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    def mapped_attributes
+      {
+        apply_application: application,
+        provider: application.provider,
+        trainee_id: raw_trainee["id"],
+        first_names: raw_trainee["first_name"],
+        last_name: raw_trainee["last_name"],
+        date_of_birth: raw_trainee["date_of_birth"],
+        email: raw_contact_details["email"],
+        # TODO: Use actual route type, if found.
```
Comment (AbigailMcP) [inline app/services/trainees/create_from_apply.rb:41]: Left over from the spike and hardcoded currently - we need to resolve. How do we know what training route they're actually on?

File: app/services/trainees/create_from_apply.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        degrees: degrees,
+      }.merge(address)
+    end
+
+    def degrees
+      raw_degrees.map { |degree| ::Degrees::CreateFromApply.call(attributes: degree) }
+    end
+
+    def address
+      raw_contact_details["country"] != "GB" ? international_address : uk_address
```
Comment (AbigailMcP) [inline app/services/trainees/create_from_apply.rb:52]: What about Northern Ireland?

**Review by AbigailMcP — COMMENTED** [review]
Some missing fields that we can maybe get now?

**Trainee**
- `trainee.subject` - looks like there's a `attributes["course"]["course_code"]`, can we use this to set the subject?

- `trainee.gender` - can we use `attributes["candidate"]["gender"]`?

- `trainee.disabilities` - can use `attributes["candidate"]["disabilities"]`? If so, do Apply disabilities map to our/DTTP values?

- `trainee.ethnic_group` - can we use `attributes["candidate"]["ethnic_group"]`? If so, do Apply ethnic groups map to our/DTTP values?

- `trainee.ethnic_background` - can we use `attributes["candidate"]["ethnic_background"]`? If so, do Apply ethnic backgrounds map to our/DTTP values? If not, can we use our `string` field `trainee.additional_ethnic_background`?

- `trainee. diversity_disclosure` - can we use the presence of any of the above values to set this enum?

**Nationalisations**
- There's a field `attributes["candidate"]["nationality"]` - can we use this to create `nationalisation`s for our trainee?

File: app/services/trainees/create_from_apply.rb (hunk 3 above)
Comment (archferns) [reply app/services/trainees/create_from_apply.rb:52]: I believe this should be okay as apply are using ISO 3166-2 :gb: 
 https://github.com/DFE-Digital/apply-for-teacher-training/blob/ca6235e6cd8449b5f24904b90449e8598c937828/config/initializers/countries.rb#L189

Comment (AbigailMcP) [reply app/services/trainees/create_from_apply.rb:52]: phew!

File: app/services/trainees/create_from_apply.rb (hunk 2 above)
Comment (gpeng) [reply app/services/trainees/create_from_apply.rb:45]: We should be able to cross reference from the courses table when we get that going properly.

File: app/services/trainees/create_from_apply.rb [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        last_name: raw_trainee["last_name"],
+        date_of_birth: raw_trainee["date_of_birth"],
+        gender: gender,
+        ethnic_group: ethnic_group,
+        ethnic_background: raw_trainee["ethnic_background"],
+        diversity_disclosure: diversity_disclosure,
+        email: raw_contact_details["email"],
+        # TODO: Use actual route type, if found.
+        training_route: 1,
+        subject: raw_course["course_code"],
```
Comment (gpeng) [inline app/services/trainees/create_from_apply.rb:47]: I don't think `course_code` is right here is it?

**Review by gpeng — COMMENTED** [review]
This is looking good. I think the values that we can't map we can handle elsewhere. We also have a lot of data we can test with from this year's applications so once we get this in we could try adding all of the providers in QA and importing them all.

File: app/lib/apply_api/code_sets/nationalities.rb [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        "slovak" => "SK",
+        "slovenian" => "SI",
+        "solomon islander" => "SB",
+        "somali" => "SO",
+        "south african" => "ZA",
+        "south korean" => "KR",
+        "south sudanese" => "SS",
+        "spanish" => "ES",
+        "sri lankan" => "LK",
+        "st helenian" => "SH",
```
Comment (archferns) [inline app/lib/apply_api/code_sets/nationalities.rb:201]: `SH` is mapped to st helenian/tristanian, which [is the correct mapping](https://en.wikipedia.org/wiki/ISO_3166-2:SH)

However, our DTTP entity mappings for [st helenian](https://github.com/DFE-Digital/register-trainee-teachers/blob/ecc21f094a666c379ebde1c6f128c387e2348f0a/app/lib/dttp/code_sets/nationalities.rb#L198) and [tristanian](https://github.com/DFE-Digital/register-trainee-teachers/blob/ecc21f094a666c379ebde1c6f128c387e2348f0a/app/lib/dttp/code_sets/nationalities.rb#L213) are different, which suggests that we might have a discrepancy.

Verified that this is the case in DTTP and updated via https://github.com/DFE-Digital/register-trainee-teachers/pull/841/commits/c0d95c4b14f640e8e688e7b020b912d9c407e9bc, however would appreciate additional eyes @gpeng

File: app/services/trainees/create_from_apply.rb (hunk 4 above)
Comment (edwardhorsford) [reply app/services/trainees/create_from_apply.rb:47]: As @gpeng  mentions, the trainee id is meant to be their own internal number - if storing an id from apply that should be kept separate.

File: app/services/degrees/create_from_apply.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    def params
+      @params ||= common_params.merge(degree_params)
+    end
+
+    def degree_params
+      uk_degree? ? uk_degree_params : non_uk_degree_params
+    end
+
+    # TODO: Check that each of degree and subject is in our
```
[the remark this answers is not in the document]
Comment (AbigailMcP) [reply app/services/degrees/create_from_apply.rb:31]: Cool - makes sense 👍

[the remark this answers is not in the document]
Comment (AbigailMcP) [reply app/services/degrees/create_from_apply.rb:31]: I'll replace the TODO with a comment along these lines.

File: app/services/trainees/create_from_apply.rb [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+    def trainee
+      @trainee ||= Trainee.new(mapped_attributes)
+    end
+
+    def mapped_attributes
+      {
+        apply_application: application,
+        provider: application.provider,
+        trainee_id: raw_trainee["id"],
```
[the remark this answers is not in the document]
Comment (AbigailMcP) [reply app/services/trainees/create_from_apply.rb:36]: ok, i'm going to remove setting them as the `trainee_id`. As we don't have a solid use-case for the apply ID yet, I'll delay adding another ID field to our trainees to store it. We can pull it out of the raw `ApplyApplication` and backfill if we need in the future.

File: app/services/degrees/create_from_apply.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        uk_degree: attributes["qualification_type"],
+        institution: institution,
+        grade: attributes["grade"],
+      }
+    end
+
+    def non_uk_degree_params
+      {
+        locale_code: Trainee.locale_codes[:non_uk],
+        # TODO: This should be the comparable_uk_degree` from Apply
```
[the remark this answers is not in the document]
Comment (AbigailMcP) [reply app/services/degrees/create_from_apply.rb:54]: `comparable_uk_degree` was added to the API, changing to this.

File: app/services/trainees/create_from_apply.rb (hunk 2 above)
Comment (AbigailMcP) [reply app/services/trainees/create_from_apply.rb:45]: Got it - pulling from the course in our table, if present.

File: app/services/trainees/create_from_apply.rb (hunk 4 above)
Comment (AbigailMcP) [reply app/services/trainees/create_from_apply.rb:47]: Changed to reference the course name in our course table, if present.

File: app/services/trainees/create_from_apply.rb [hunk 9]
```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module Trainees
+  class CreateFromApply
```
[the remark this answers is not in the document]
Comment (AbigailMcP) [reply app/services/trainees/create_from_apply.rb:4]: @gpeng - what are your thoughts re this? Should we run this service and create trainees off the back of every `ApplyApi::ImportApplicationsJob`, if any are newly imported?

Also worth noting that this job does `ApplyApplication.find_or_initialize_by(apply_id: apply_id)`. If an `ApplyApplication` is updated, do we update the trainee? I'm assuming not...

[the remark this answers is not in the document]
Comment (gpeng) [reply app/services/trainees/create_from_apply.rb:4]: We were going to leave it not being run for now until the UR has finished and we've decided how/when we want to import them. I don't think we want to update the trainee if the application has been updated so we should only import those that have not already been imported. We should probably defer all that for a separate ticket so as we can give it a bit of thought.

File: app/lib/apply_api/code_sets/nationalities.rb (hunk 5 above)
Comment (gpeng) [reply app/lib/apply_api/code_sets/nationalities.rb:201]: The mapping we have for DTTP is incorrect. They should both be - `77c297b7-5c62-e711-80d1-005056ac45bb`

I've updated in https://github.com/DFE-Digital/register-trainee-teachers/pull/863

**Review by gpeng — APPROVED** [review]
I think we should look to merge this and then address any of the smaller things along the way if needed.

File: app/services/trainees/create_from_apply.rb [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    def nationalities
+      @nationalities ||= Nationality.where(name: nationality_names)
+    end
+
+    def nationality_names
+      raw_trainee["nationality"].map { |nationality| ApplyApi::CodeSets::Nationalities::MAPPING[nationality] }
+    end
+
+    def course
+      @course ||= Course.find_by(code: raw_course["course_code"])
```
Comment (defong) [inline app/services/trainees/create_from_apply.rb:115]: We need `provider` or `provider.code`

Comment (AbigailMcP) [reply app/services/trainees/create_from_apply.rb]: thanks!

**Pull request merged**